Add torch ops for d2go models

dncnbuck opened the pull request:
added torch ops
```
nms
repeat_interleave
numel
roi_align
logicaland
```

ops with minor patches

```
clamp
narrow 
index - handle broadcasting indicies
max -  inputs are len == 1
split - handle case when num_splits == 1
to -  handle the case where the dtype is not set, this should be inferred from the Tensor dtype. see, https://pytorch.org/docs/stable/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to
tupleunpack - handle case when len(output) == 1
```

these were added in order to make progress with a conversion of 
 pytorch Mask-RCNN model from d2go / detectron2.

Form more details see:  https://github.com/apple/coremltools/issues/1505#issuecomment-1148436445

TobyRoseman commented on coremltools/converters/mil/frontend/torch/ops.py:
```diff
@@ -3357,6 +3384,13 @@ def to(context, node):
+    if dtype is None:
+        out = mb.identity(x=_input, name=node.name)
+        context.add(out, node.name)
+        return = 0 # TODO: infer from Tensor (spoiler in this case we care about its f32 => 6)
```
This might make sense for your model but I don't think this makes sense in general. Probably want to do the TODO or make the body of this if-statement raise an exception.

dncnbuck replied:
My bad, looks like a bad typo there and I left in a TODO 😵‍💫. 

This should be 
```
if dtype is None:
        out = mb.identity(x=_input, name=node.name)
        context.add(out, node.name)
        return
```
that is if the dtype to convert the tensor to is None, then  the inferred dtype is that of the input tensor. Here I then just make a copy of the input tensor and return early. I think this is aligned with the torch docs linked (https://pytorch.org/docs/stable/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to)

Does that make sense?

TobyRoseman replied:
That makes sense.

TobyRoseman reviewed (commented):
Thanks for the PR. These changes look good. I just had a couple of questions.

Before we can merge it. We need unit tests for the bug fixes and the new ops. @dncnbuck - can you add that?

It also looks like there is a merge conflict.

fukatani commented on coremltools/converters/mil/frontend/torch/ops.py:
```diff
@@ -4357,3 +4408,162 @@ def scatter_add(context, node):
+    context.add(out, torch_name=node.name)
+
+@register_torch_op(torch_alias=["__and_", '__and__'])
+def logicaland(context, node):
```
`logical_and` was already implemented.
We need this function?

dncnbuck replied:
@fukatani Awesome! Thanks, I've removed the op and used the logical_and version  and added the torch alias there.

A note on the alias `__and_`. Within the model the RPN requires the op
```
item = torch.__and__(torch.gt(widths0, 0.), torch.gt(heights0, 0.))
```
However `convert_nodes` assumes that ops ending in `_` are inplace ops and removes the trailing `_` before op lookup. Maybe I should just patch this issue instead and leave only the alias `__and__`.  yeah I think i'll do this

TobyRoseman replied:
Making this change as a separate PR sounds good.

Could someone provide a toy network with a `__and__` layer?

dncnbuck commented on coremltools/converters/mil/frontend/torch/ops.py:
```diff
@@ -4357,3 +4408,162 @@ def scatter_add(context, node):
+@register_torch_op()
+def roi_align(context, node):
+    """
+    https://github.com/apple/coremltools/blob/655b3be5cc0d42c3c4fa49f0f0e4a93a26b3e492/mlmodel/format/NeuralNetwork.proto#L2239
```
This was personal documentation link to the coreml NN crop_and_resize layer, which I was referencing to ensure the input specs were correctly implemented. I've removed now.

dncnbuck commented:
@TobyRoseman "We need unit tests for the bug fixes and the new ops. @dncnbuck - can you add that?"

Yeah good point. I'll be able to add these but this might not happen today sadly. I'll try to get to it this week 😄

TobyRoseman commented:
> I'll be able to add these but this might not happen today sadly. I'll try to get to it this week 😄

No worries about the timing. Let me know if you have any questions when adding the unit tests.

ajsharda17 reviewed (changes requested):
I think you should update to the latest code, as I don't think this can be merged atm without the update to the latest code.

dncnbuck commented:
> > I'll be able to add these but this might not happen today sadly. I'll try to get to it this week 😄
> 
> No worries about the timing. Let me know if you have any questions when adding the unit tests.

Hi @TobyRoseman I've got some time to add the unit tests now, but I'm running into issues testing the coremltools suite. I am receiving test exceptions when testing the main branch of the lib. I assume my dev setup is not currently setup correctly.

```
E           TypeError: predict(): incompatible function arguments. The following argument types are supported:
E               1. (self: coremltools.libcoremlpython._MLModelProxy, arg0: dict, arg1: bool) -> dict
E           
```

luliuzee commented:
@dncnbuck did you manage to fix this issue by any chance? I will be thankful for your response

dncnbuck commented:
@TobyRoseman sorry this dropped off of my radar these last months. I've updated the Pr to include some tests for ops and removed the repeat interleave op implementation for the time being as it was not implemented well enough so I will hold out on adding that until I can give it some time

TobyRoseman commented on coremltools/converters/mil/frontend/torch/ops.py:
```diff
@@ -3398,6 +3405,18 @@ def _slice(context, node):
     context.add(res)
 
 
+def _num_splits_and_sizes(split_sizes):
```
Should this just be an inner method of the `split` method?

TobyRoseman commented on coremltools/converters/mil/frontend/torch/ops.py:
```diff
@@ -4501,8 +4551,106 @@ def scatter(context, node):
     _scatter(context, inputs, 'add', node.name)
+
+@register_torch_op
+def roi_align(context, node):
```
Are there unit tests for this method?

TobyRoseman commented:
@dncnbuck - no worries about this falling off your radar. Thanks for getting back to it. 

Please rebase this pull request on top of `main`. Then I will kick off a CI run.

diegogranziol commented:
Has this PR been included? i.e. is it possible to convert d2go models to coreml?
